File: ECMAScript/Sorts/HeapSort.js
/*
Brian Taylor Vann
github.com/taylor-vann

Description:
- Function for heap sort

Requirements:
- None
*/


const heapSort = (arr) => {
  l = arr.length;
  h = Math.floor(l / 2);

  if(l < 2) {
    return arr;
  }

  // from last parent to first, sift down so largest value on top
  while(h > -1) {
    siftDown(arr, h, l);
    h -= 1;
  }

  // walk back and swap first and index and sift new first down
  for(var j = l - 1; j > -1; j--) {
    var tmp = arr[0];
    arr[0] = arr[j];
    arr[j] = tmp;

    siftDown(arr, 0, j);
  }

  return arr;
}


function siftDown(arr, i, l) {
  var c = i;
  var n = 2 * c + 1;

  while(n < l) {
    if(n < l - 1 && arr[n] < arr[n + 1]) {
      n += 1;
    }

    if(arr[c] < arr[n]) {
      var tmp = arr[c];
      arr[c] = arr[n];
      arr[n] = tmp;

      c = n;
      n = c * 2 + 1;
    }else{
      break
    }
  }
}

console.log(heapSort([]));
console.log(heapSort([1]));
console.log(heapSort([1, 3, 5]));
console.log(heapSort([4, 8, 6, 9, 0, 1, 10]));
console.log(heapSort([11, 7, 8, 7, 0]));

File: Python3/Sorts/HeapSort.py
"""
Brian Taylor Vann
github.com/taylor-vann
"""

def heap_sort(arr):
    if len(arr) < 2:
        return arr

    for j in range(len(arr) // 2, -1, -1):
        sift_down(arr, j, len(arr))

    for k in range(len(arr) - 1, -1, -1):
        arr[0], arr[k] = arr[k], arr[0]
        sift_down(arr, 0, k)

    return arr


def sift_down(arr, first, last):
    curr = first
    nxt = 2 * curr + 1

    while nxt < last:
        if nxt < last - 1 and arr[nxt] < arr[nxt + 1]:
            nxt += 1

        if arr[curr] > arr[nxt]:
            break

        arr[curr], arr[nxt] = arr[nxt], arr[curr]
        curr, nxt = nxt, curr * 2 + 1